core/src/main/java/jenkins/tasks/filters/impl/RetainVariablesLocalRule.java

```diff
@@ -0,0 +1,211 @@
+import hudson.EnvVars;
+import hudson.Extension;
+import hudson.util.FormValidation;
+import jenkins.tasks.filters.EnvVarsFilterRuleContext;
+import jenkins.tasks.filters.EnvVarsFilterLocalRule;
+import jenkins.tasks.filters.EnvVarsFilterLocalRuleDescriptor;
+import jenkins.tasks.filters.EnvVarsFilterableBuilder;
+import org.apache.commons.lang.StringUtils;
+import org.jenkinsci.Symbol;
+import org.jvnet.localizer.Localizable;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.DataBoundSetter;
+import org.kohsuke.stapler.QueryParameter;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+
+/**
+ * Local rule that removes all the non-retained variables for that step.
+ *
+ * @since TODO
+ */
+@Restricted(NoExternalUse.class)
+public class RetainVariablesLocalRule implements EnvVarsFilterLocalRule {
+
+    /**
+     * List of lowercase names of variable that will be retained from removal
+     */
+    private String variables = "";
+    private boolean retainCharacteristicEnvVars = true;
+    private ProcessVariablesHandling processVariablesHandling = ProcessVariablesHandling.RESET;
+
+    @DataBoundConstructor
+    public RetainVariablesLocalRule() {
+    }
+
+    @DataBoundSetter
+    public void setVariables(@NonNull String variables) {
+        this.variables = variables;
+    }
+
+    private static List<String> convertStringToList(@NonNull String variablesCommaSeparated) {
+        String[] variablesArray = variablesCommaSeparated.split("\\s+");
+        List<String> variables = new ArrayList<>();
+        for (String nameFragment : variablesArray) {
+            if (StringUtils.isNotBlank(nameFragment)) {
+                variables.add(nameFragment.toLowerCase(Locale.ENGLISH));
+            }
+        }
+
+        Collections.sort(variables); // TODO do we really want to sort this?
+        return variables;
+    }
+
+    // for jelly view
+    @Restricted(NoExternalUse.class)
+    public @NonNull String getVariables() {
+        return variables;
+    }
+
+    @DataBoundSetter
+    public void setRetainCharacteristicEnvVars(boolean retainCharacteristicEnvVars) {
+        this.retainCharacteristicEnvVars = retainCharacteristicEnvVars;
+    }
+
+    public boolean isRetainCharacteristicEnvVars() {
+        return retainCharacteristicEnvVars;
+    }
+
+    private List<String> variablesToRetain() {
+        List<String> vars = new ArrayList<>(convertStringToList(this.variables));
+        if (isRetainCharacteristicEnvVars()) {
+            // TODO Make the 'HUDSON_COOKIE' variable less special so we can remove it.
+            vars.addAll(Arrays.asList("jenkins_server_cookie", "hudson_server_cookie", "job_name", "job_base_name", "build_number", "build_id", "build_tag"));
```

Comment:
As the term "characteristic" isn't widely used or known it would be helpful to clarify it a bit. I suggest pulling out these strings into a static field with a clarifying name and maybe a comment.